src/Microsoft.AspNet.Razor/SourceLocation.cs

```diff
@@ -112,33 +164,43 @@ private static SourceLocation CreateUndefined()
             return sl;
         }
 
-        public static bool operator <(SourceLocation left, SourceLocation right)
-        {
-            return left.CompareTo(right) < 0;
-        }
-
-        public static bool operator >(SourceLocation left, SourceLocation right)
-        {
-            return left.CompareTo(right) > 0;
-        }
-
-        public static bool operator ==(SourceLocation left, SourceLocation right)
-        {
-            return left.Equals(right);
-        }
-
-        public static bool operator !=(SourceLocation left, SourceLocation right)
-        {
-            return !left.Equals(right);
-        }
-
+        /// <summary>
+        /// Adds two <see cref="SourceLocation"/>s.
+        /// </summary>
+        /// <param name="left">The left operand.</param>
+        /// <param name="right">The right operand.</param>
+        /// <returns>A <see cref="SourceLocation"/> that is the sum of the left and right operands.</returns>
+        /// <exception cref="ArgumentException">if the <see cref="FilePath"/> of the left and right operands
+        /// are different.</exception>
         public static SourceLocation operator +(SourceLocation left, SourceLocation right)
         {
+            if (!string.Equals(left.FilePath, right.FilePath, StringComparison.OrdinalIgnoreCase))
+            {
+                throw new ArgumentException(
+                    RazorResources.FormatSourceLocationFilePathDoesNotMatch(nameof(SourceLocation)),
+                    nameof(right));
+            }
+
             return Add(left, right);
         }
 
+        /// <summary>
+        /// Subtracts two <see cref="SourceLocation"/>s.
+        /// </summary>
+        /// <param name="left">The left operand.</param>
+        /// <param name="right">The right operand.</param>
+        /// <returns>A <see cref="SourceLocation"/> that is the difference of the left and right operands.</returns>
+        /// <exception cref="ArgumentException">if the <see cref="FilePath"/> of the left and right operands
+        /// are different.</exception>
         public static SourceLocation operator -(SourceLocation left, SourceLocation right)
         {
+            if (!string.Equals(left.FilePath, right.FilePath, StringComparison.OrdinalIgnoreCase))
+            {
+                throw new ArgumentException(
```

Comment:
I think we generally avoid throwing the exception from `Ensure` methods to ensure we get the most accurate stack traces.